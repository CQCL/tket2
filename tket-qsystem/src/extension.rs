//! This module defines the Hugr extensions used by tket-qsystem.
pub mod futures;
pub mod qsystem;
pub mod random;
pub mod result;
pub mod utils;
pub mod wasm;
