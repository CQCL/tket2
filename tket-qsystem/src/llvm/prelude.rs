//! LLVM lowering implementations for HUGR prelude operations.

use std::hash::{DefaultHasher, Hash as _, Hasher as _};
use tket::hugr::{self, llvm::inkwell};

use anyhow::{anyhow, ensure, Result};
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::extension::prelude::PreludeCodegen;
use hugr::llvm::types::TypingSession;
use inkwell::attributes::{Attribute, AttributeLoc};
use inkwell::module::Linkage;
use inkwell::types::{BasicType, IntType};
use inkwell::values::{BasicValue, BasicValueEnum, FunctionValue, StructValue};
use inkwell::AddressSpace;
use tket::hugr::{HugrView, Node};

#[derive(Clone)]
/// Codegen extension for the QIS prelude.
pub struct QISPreludeCodegen;

impl PreludeCodegen for QISPreludeCodegen {
    fn usize_type<'c>(&self, session: &TypingSession<'c, '_>) -> IntType<'c> {
        session.iw_context().i64_type()
    }

    /// Return the llvm type of [`tket::hugr::extension::prelude::qb_t`].
    fn qubit_type<'c>(&self, session: &TypingSession<'c, '_>) -> impl BasicType<'c> {
        session.iw_context().i64_type()
    }

    /// A panic ends all shots, this is done by using a return code > 1000.
    /// The lowering of the panic HUGR op shifts the return code held by the op by 1000.
    /// So valid values for `signal` in the panic HUGR op are 1-(2^31 - 1001).
    fn emit_panic<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_panic: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-(2^31 - 1001), otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let return_code = {
            let shift = ctx.iw_context().i32_type().const_int(1000, false);
            ctx.builder()
                .build_int_add(return_code.into_int_value(), shift, "shift_code")?
        };

        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    /// An exit ends the current shot using a return code `rc`, where `1 <= rc <= 1000`.
    fn emit_exit<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_exit: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-1000, otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    fn emit_print<H: HugrView<Node = Node>>(
        &self,
        _ctx: &mut EmitFuncContext<H>,
        _text: inkwell::values::BasicValueEnum,
    ) -> Result<()> {
        Ok(())
    }

    fn emit_const_error<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        err: &hugr::extension::prelude::ConstError,
    ) -> Result<BasicValueEnum<'c>> {
        let err_ty = ctx
            .llvm_type(&hugr::extension::prelude::error_type())?
            .into_struct_type();
        let signal = err_ty
            .get_field_type_at_index(0)
            .unwrap()
            .into_int_type()
            .const_int(u64::from(err.signal), false);

        let message = emit_global_string(ctx, &err.message, "e_", "EXIT:INT:")?;
        let err = err_ty.const_named_struct(&[signal.into(), message]);
        Ok(err.into())
    }

    fn emit_const_string<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        str: &hugr::extension::prelude::ConstString,
    ) -> Result<BasicValueEnum<'c>> {
        emit_global_string(ctx, str.value(), "s_", "")
    }
}

impl QISPreludeCodegen {
    fn get_panic<'c, H: HugrView<Node = Node>>(
        ctx: &EmitFuncContext<'c, '_, H>,
    ) -> Result<FunctionValue<'c>> {
        let iwc = ctx.iw_context();
        let func_type = iwc.void_type().fn_type(
            &[
                iwc.i32_type().into(),
                iwc.i8_type().ptr_type(AddressSpace::default()).into(),
            ],
            false,
        );
        let f = ctx.get_extern_func("panic", func_type)?;
        let noreturn = Attribute::get_named_enum_kind_id("noreturn");
        debug_assert!(noreturn != 0);
        f.add_attribute(
            AttributeLoc::Function,
            iwc.create_enum_attribute(noreturn, 0),
        );

        Ok(f)
    }
}

/// Emit a global string constant with a unique name.
///
/// The symbol name is constructed with `symbol_prefix` and ``str`.
///
/// The string is prefixed with `str_prefix`.
pub fn emit_global_string<'c, H: HugrView<Node = Node>>(
    ctx: &EmitFuncContext<'c, '_, H>,
    str: impl AsRef<str>,
    symbol_prefix: impl AsRef<str>,
    str_prefix: impl AsRef<str>,
) -> Result<BasicValueEnum<'c>> {
    let (str, symbol_prefix, str_prefix) =
        (str.as_ref(), symbol_prefix.as_ref(), str_prefix.as_ref());
    let tagged_str = format!("{str_prefix}{str}");
    let tagged_str_bytes = tagged_str.as_bytes();
    let tagged_str_len = tagged_str_bytes.len();
    ensure!(
        tagged_str_len < 256,
        "Constant string too long: {tagged_str_len} >= 256"
    );

    let tgt_str_bytes = [&[u8::try_from(tagged_str_len)?], tagged_str_bytes].concat();
    let tgt_str_value = ctx.iw_context().const_string(&tgt_str_bytes, false);

    let global = {
        // we need a unique name for the global. We use a prefix(of length
        // 10) of the string, followed by the hash of the whole string.
        //
        // This should be enough, but for absolute guarantee of success we
        // append ".i", i an integer and find an available name.
        let hash = {
            let mut s = DefaultHasher::default();
            tgt_str_bytes.hash(&mut s);
            let h = s.finish();
            let lower32 = h as u32;
            format!("{lower32:08X}")
        };
        // Use unicode-aware character slicing to prevent slicing through multi-byte characters
        let prefix_chars: String = str.chars().take(10).collect();
        let name_prefix = format!("{symbol_prefix}{prefix_chars}.{hash}");

        (0..)
            .find_map(|i| {
                let name = format!("{name_prefix}.{i}");
                if let Some(global) = ctx.get_current_module().get_global(&name) {
                    if let Some(v) = global.get_initializer() {
                        if v == tgt_str_value {
                            // We've found a global with the same value, use that.
                            return Some(global);
                        }
                    }
                } else {
                    let global =
                        ctx.get_current_module()
                            .add_global(tgt_str_value.get_type(), None, &name);
                    global.set_initializer(&tgt_str_value);
                    global.set_linkage(Linkage::Private);
                    global.set_constant(true);
                    return Some(global);
                }
                None
            })
            .unwrap()
    };
    Ok(ctx
        .builder()
        .build_pointer_cast(
            global.as_pointer_value(),
            ctx.iw_context().i8_type().ptr_type(AddressSpace::default()),
            "",
        )?
        .as_basic_value_enum())
}

#[cfg(test)]
mod test {
    use hugr::builder::{Dataflow, DataflowHugr};
    use hugr::extension::prelude::{self, qb_t, ConstError, EXIT_OP_ID, PANIC_OP_ID};
    use hugr::extension::PRELUDE;
    use hugr::llvm::check_emission;
    use hugr::llvm::emit::test::SimpleHugrConfig;

    use hugr::llvm::test::llvm_ctx;
    use hugr::types::TypeArg;
    use rstest::rstest;

    use super::*;

    #[rstest]
    fn test_panic_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        // Create a hugr that has a panic message
        let error_val = ConstError::new(42, "PANIC");
        let type_arg_q = TypeArg::Runtime(qb_t());
        let type_arg_2q = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let panic_op = PRELUDE
            .instantiate_extension_op(&PANIC_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(panic_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }

    #[rstest]
    fn test_exit_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        let error_val = ConstError::new(42, "EXIT");
        let type_arg_q: TypeArg = TypeArg::Runtime(qb_t());
        let type_arg_2q: TypeArg = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let exit_op = PRELUDE
            .instantiate_extension_op(&EXIT_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(exit_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }
}
