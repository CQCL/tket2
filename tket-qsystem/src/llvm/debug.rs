//! LLVM lowering implementations for "tket.debug" extension.

use crate::llvm::prelude::emit_global_string;
use anyhow::{anyhow, bail, Result};
use hugr::extension::simple_op::MakeExtensionOp;
use hugr::llvm::custom::CodegenExtension;
use hugr::llvm::emit::{EmitFuncContext, EmitOpArgs};
use hugr::llvm::inkwell::AddressSpace;
use hugr::llvm::CodegenExtsBuilder;
use hugr::ops::ExtensionOp;
use hugr::{HugrView, Node};
use tket::extension::debug::{StateResult, DEBUG_EXTENSION_ID, STATE_RESULT_OP_ID};

use super::array_utils::{struct_1d_arr_alloc, struct_1d_arr_ptr_t, ArrayLowering, ElemType};

static TAG_PREFIX: &str = "USER:";

/// Codegen extension for debug functionality.
#[derive(Default)]
pub struct DebugCodegenExtension<AL: ArrayLowering> {
    array_lowering: AL,
}

impl<AL: ArrayLowering> DebugCodegenExtension<AL> {
    /// Creates a new [DebugCodegenExtension] with specified array lowering.
    pub const fn new(array_lowering: AL) -> Self {
        Self { array_lowering }
    }
}

impl<AL: ArrayLowering> CodegenExtension for DebugCodegenExtension<AL> {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: CodegenExtsBuilder<'a, H>,
    ) -> CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder.extension_op(DEBUG_EXTENSION_ID, STATE_RESULT_OP_ID, {
            move |context, args| self.emit_state_result(context, args)
        })
    }
}

impl<AL: ArrayLowering> DebugCodegenExtension<AL> {
    /// Lower the `debug` `StateResult` op.
    fn emit_state_result<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
    ) -> Result<()> {
        let builder = ctx.builder();

        // Types (qubits are just i64s).
        let iw_ctx = ctx.iw_context();
        let void_t = iw_ctx.void_type();
        let i8_ptr_t = iw_ctx.i8_type().ptr_type(AddressSpace::default());
        let i64_t = iw_ctx.i64_type();

        // Tag arguments.
        let state_result = StateResult::from_extension_op(args.node().as_ref())?;
        let tag = state_result.tag;
        if tag.is_empty() {
            bail!("Empty state result tag received");
        }
        let tag_ptr = emit_global_string(ctx, tag, "res_", format!("{TAG_PREFIX}STATE:"))?;
        let tag_len = {
            let mut l = builder
                .build_load(tag_ptr.into_pointer_value(), "tag_len")?
                .into_int_value();
            if l.get_type() != i64_t {
                l = builder.build_int_z_extend(l, i64_t, "tag_len")?;
            }
            l
        };

        // Qubit array argument.
        let array_len = state_result.num_qubits;
        let [qubits] = args
            .inputs
            .try_into()
            .map_err(|_| anyhow!(format!("StateResult expects a qubit array argument")))?;
        let qubits_array = self.array_lowering.array_to_ptr(builder, qubits)?;
        let (qubits_ptr, _) = struct_1d_arr_alloc(
            iw_ctx,
            builder,
            array_len.try_into()?,
            &ElemType::Int,
            qubits_array,
        )?;

        // Build the function call.
        let fn_state_result = ctx.get_extern_func(
            "print_state_result",
            void_t.fn_type(
                &[
                    i8_ptr_t.into(),
                    i64_t.into(),
                    struct_1d_arr_ptr_t(iw_ctx, &ElemType::Int).into(),
                ],
                false,
            ),
        )?;

        builder.build_call(
            fn_state_result,
            &[tag_ptr.into(), tag_len.into(), qubits_ptr.into()],
            "print_state_result",
        )?;
        args.outputs.finish(builder, [qubits])
    }
}

#[cfg(test)]
#[allow(deprecated)] // TODO remove after switching to new array lowering
mod test {
    use tket::extension::debug::StateResult;

    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::llvm::check_emission;
    use hugr::llvm::test::llvm_ctx;
    use hugr::llvm::test::single_op_hugr;
    use hugr::llvm::test::TestContext;

    use crate::llvm::array_utils::DEFAULT_HEAP_ARRAY_LOWERING;
    use crate::llvm::array_utils::DEFAULT_STACK_ARRAY_LOWERING;
    use crate::llvm::prelude::QISPreludeCodegen;

    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case::state_result(1, StateResult::new("test_state_result".to_string(), 2), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::state_result(2, StateResult::new("test_state_result".to_string(), 2), &DEFAULT_HEAP_ARRAY_LOWERING)]
    fn emit_debug_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: StateResult,
        #[case] array_lowering: &'static (impl ArrayLowering + Clone),
    ) {
        let pcg = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| {
            ceb.add_extension(DebugCodegenExtension::new(array_lowering.clone()))
                .add_extension(array_lowering.codegen_extension())
                .add_prelude_extensions(pcg.clone())
                .add_default_int_extensions()
                .add_float_extensions()
        });
        let ext_op = op.to_extension_op().unwrap().into();
        let hugr = single_op_hugr(ext_op);
        check_emission!(hugr, llvm_ctx);
    }
}
