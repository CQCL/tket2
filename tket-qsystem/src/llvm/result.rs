//! LLVM lowering implementations for "tket.results" extension.

use crate::extension::result::{ResultArgs, ResultOp, ResultOpDef};
use crate::llvm::prelude::emit_global_string;
use anyhow::{anyhow, bail, Result};
use hugr::llvm::custom::CodegenExtension;
use hugr::llvm::emit::{EmitFuncContext, EmitOpArgs};
use hugr::llvm::inkwell;
use hugr::llvm::sum::LLVMSumValue;
use hugr::llvm::types::HugrSumType;
use hugr::llvm::CodegenExtsBuilder;
use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::types::{FloatType, IntType, PointerType, VoidType};
use inkwell::values::{BasicValueEnum, FunctionValue, IntValue};
use inkwell::AddressSpace;
use tket::hugr::extension::simple_op::MakeExtensionOp;
use tket::hugr::ops::ExtensionOp;
use tket::hugr::{HugrView, Node};

use super::array_utils::{struct_1d_arr_alloc, struct_1d_arr_ptr_t, ArrayLowering, ElemType};

static TAG_PREFIX: &str = "USER:";

/// Codegen extension for results
#[derive(Default)]
pub struct ResultsCodegenExtension<AL: ArrayLowering> {
    array_lowering: AL,
}

impl<AL: ArrayLowering> ResultsCodegenExtension<AL> {
    /// Creates a new [ResultsCodegenExtension] with specified array lowering.
    pub const fn new(array_lowering: AL) -> Self {
        Self { array_lowering }
    }
}

impl<AL: ArrayLowering + Clone> CodegenExtension for ResultsCodegenExtension<AL> {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: CodegenExtsBuilder<'a, H>,
    ) -> CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder.simple_extension_op::<ResultOpDef>(move |context, args, _op| {
            let op = ResultOp::from_extension_op(args.node().as_ref())?;
            ResultEmitter(context, self.array_lowering.clone()).emit(args, &op)
        })
    }
}

struct ResultEmitter<'c, 'd, 'e, H: HugrView<Node = Node>, AL: ArrayLowering>(
    &'d mut EmitFuncContext<'c, 'e, H>,
    AL,
);

impl<'c, H: HugrView<Node = Node>, AL: ArrayLowering + Clone> ResultEmitter<'c, '_, '_, H, AL> {
    fn iw_context(&self) -> &'c Context {
        self.0.typing_session().iw_context()
    }

    fn int_t(&self) -> IntType<'c> {
        self.iw_context().i64_type()
    }

    fn float_t(&self) -> FloatType<'c> {
        self.iw_context().f64_type()
    }

    fn bool_t(&self) -> IntType<'c> {
        self.iw_context().bool_type()
    }

    fn i8_ptr_t(&self) -> PointerType<'c> {
        self.iw_context()
            .i8_type()
            .ptr_type(AddressSpace::default())
    }

    fn void_t(&self) -> VoidType<'c> {
        self.iw_context().void_type()
    }

    fn builder(&self) -> &Builder<'c> {
        self.0.builder()
    }

    fn get_func_print(&self, op: &ResultOp) -> Result<FunctionValue<'c>> {
        // The first two parameters are the same for all print function variants
        let mut params = vec![self.i8_ptr_t().into(), self.int_t().into()];
        let symbol = match op.result_op {
            ResultOpDef::Bool => {
                params.push(self.bool_t().into());
                "print_bool"
            }
            ResultOpDef::Int => {
                params.push(self.int_t().into());
                "print_int"
            }
            ResultOpDef::UInt => {
                params.push(self.int_t().into());
                "print_uint"
            }
            ResultOpDef::F64 => {
                params.push(self.float_t().into());
                "print_float"
            }
            ResultOpDef::ArrBool => {
                params.push(struct_1d_arr_ptr_t(self.iw_context(), &ElemType::Bool).into());
                "print_bool_arr"
            }
            ResultOpDef::ArrInt => {
                params.push(struct_1d_arr_ptr_t(self.iw_context(), &ElemType::Int).into());
                "print_int_arr"
            }
            ResultOpDef::ArrUInt => {
                params.push(struct_1d_arr_ptr_t(self.iw_context(), &ElemType::Uint).into());
                "print_uint_arr"
            }
            ResultOpDef::ArrF64 => {
                params.push(struct_1d_arr_ptr_t(self.iw_context(), &ElemType::Float).into());
                "print_float_arr"
            }
        };

        let func_t = self.void_t().fn_type(&params, false);
        self.0.get_extern_func(symbol, func_t)
    }

    fn generate_global_tag(
        &self,
        args: &EmitOpArgs<'c, '_, ExtensionOp, H>,
        type_tag: impl AsRef<str>,
    ) -> Result<(BasicValueEnum<'_>, IntValue<'_>)> {
        let type_tag = type_tag.as_ref();
        let result_op = ResultOp::from_extension_op(args.node().as_ref())?;

        let tag = result_op.tag;
        if tag.is_empty() {
            bail!("Empty result tag received");
        }

        let tag_ptr = emit_global_string(self.0, tag, "res_", format!("{TAG_PREFIX}{type_tag}"))?;
        let tag_len = {
            let mut l = self
                .0
                .builder()
                .build_load(tag_ptr.into_pointer_value(), "tag_len")?
                .into_int_value();
            if self.int_t() != l.get_type() {
                l = self
                    .builder()
                    .build_int_z_extend(l, self.int_t(), "tag_len")?;
            }
            l
        };

        Ok((tag_ptr, tag_len))
    }

    fn build_print_array_call(
        &self,
        val: BasicValueEnum,
        op: &ResultOp,
        data_type: &ElemType,
        tag_ptr: BasicValueEnum,
        tag_len: IntValue,
    ) -> Result<()> {
        // TODO update to return array after https://github.com/CQCL/tket2/issues/922
        let ResultArgs::Array(_, length) = op.args else {
            bail!("Expected array argument")
        };

        let print_fn = self.get_func_print(op)?;
        let array = self.1.array_to_ptr(self.builder(), val)?;
        let (array_ptr, _) = struct_1d_arr_alloc(
            self.iw_context(),
            self.builder(),
            length.try_into()?,
            data_type,
            array,
        )?;
        self.builder().build_call(
            print_fn,
            &[tag_ptr.into(), tag_len.into(), array_ptr.into()],
            "",
        )?;
        Result::Ok(())
    }

    /// Function to help lower the tket result extension.
    fn emit(&self, args: EmitOpArgs<'c, '_, ExtensionOp, H>, op: &ResultOp) -> Result<()> {
        let print_fn = self.get_func_print(op)?;
        match op.result_op {
            ResultOpDef::Bool => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "BOOL:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let bool_type = self.0.llvm_sum_type(HugrSumType::new_unary(2))?;
                let val = LLVMSumValue::try_new(val, bool_type)
                    .map_err(|_| anyhow!("bool_type expects a value"))?
                    .build_get_tag(self.builder())?;
                let b_type = self.bool_t();
                let trunc_val = self.builder().build_int_truncate(val, b_type, "")?;
                self.builder().build_call(
                    print_fn,
                    &[tag_ptr.into(), tag_len.into(), trunc_val.into()],
                    "print_bool",
                )?;
            }
            ResultOpDef::Int => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "INT:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_int expects one input"))?;
                self.builder().build_call(
                    print_fn,
                    &[tag_ptr.into(), tag_len.into(), val.into()],
                    "print_int",
                )?;
            }
            ResultOpDef::UInt => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "INT:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_uint expects one input"))?;
                self.builder().build_call(
                    print_fn,
                    &[tag_ptr.into(), tag_len.into(), val.into()],
                    "print_uint",
                )?;
            }
            ResultOpDef::F64 => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "FLOAT:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_float expects one input"))?;
                self.builder().build_call(
                    print_fn,
                    &[tag_ptr.into(), tag_len.into(), val.into()],
                    "print_int",
                )?;
            }
            ResultOpDef::ArrBool => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "BOOLARR:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_arr_bool expects one input"))?;
                self.build_print_array_call(val, op, &ElemType::Bool, tag_ptr, tag_len)?;
            }
            ResultOpDef::ArrInt => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "INTARR:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_arr_int expects one input"))?;
                self.build_print_array_call(val, op, &ElemType::Int, tag_ptr, tag_len)?;
            }
            ResultOpDef::ArrUInt => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "INTARR:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_arr_uint expects one input"))?;
                self.build_print_array_call(val, op, &ElemType::Uint, tag_ptr, tag_len)?;
            }
            ResultOpDef::ArrF64 => {
                let (tag_ptr, tag_len) = self.generate_global_tag(&args, "FLOATARR:").unwrap();
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_arr_float expects one input"))?;
                self.build_print_array_call(val, op, &ElemType::Float, tag_ptr, tag_len)?;
            }
        }
        args.outputs.finish(self.builder(), [])
    }
}

#[cfg(test)]
#[allow(deprecated)] // TODO remove after switching to new array lowering
mod test {
    use crate::extension::result::ResultOp;
    use crate::llvm::array_utils::DEFAULT_HEAP_ARRAY_LOWERING;
    use crate::llvm::array_utils::DEFAULT_STACK_ARRAY_LOWERING;

    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::llvm::check_emission;
    use hugr::llvm::test::llvm_ctx;
    use hugr::llvm::test::single_op_hugr;
    use hugr::llvm::test::TestContext;

    use crate::llvm::prelude::QISPreludeCodegen;

    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case::bool(1, ResultOp::new_bool("test_bool"), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::int(2, ResultOp::new_int("test_int", 6), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::uint(3, ResultOp::new_uint("test_uint", 6), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::f64(4, ResultOp::new_f64("test_f64"), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::arr_bool(5, ResultOp::new_bool("test_arr_bool").array_op(10), &DEFAULT_HEAP_ARRAY_LOWERING)]
    #[case::arr_bool(6, ResultOp::new_bool("test_arr_bool").array_op(10), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::arr_int(7, ResultOp::new_int("test_arr_int", 6).array_op(10), &DEFAULT_HEAP_ARRAY_LOWERING)]
    #[case::arr_int(8, ResultOp::new_int("test_arr_int", 6).array_op(10), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::arr_uint(9, ResultOp::new_uint("test_arr_uint", 6).array_op(10), &DEFAULT_HEAP_ARRAY_LOWERING)]
    #[case::arr_int(10, ResultOp::new_int("test_arr_int", 6).array_op(10), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::arr_f64(11, ResultOp::new_f64("test_arr_f64").array_op(10), &DEFAULT_HEAP_ARRAY_LOWERING)]
    // test cases for various tags
    #[case::unicode_tag(12, ResultOp::new_int("测试字符串", 6), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::special_chars(13, ResultOp::new_uint("test!@#$%^&*()", 6), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[should_panic(expected = "Constant string too long")]
    #[case::very_long_tag(14, ResultOp::new_f64("x".repeat(256)), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::whitespace(15, ResultOp::new_bool("   spaces   tabs\t\t\tnewlines\n\n\n"), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::emoji(16, ResultOp::new_bool("🚀👨‍👩‍👧‍👦🌍"), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[should_panic(expected = "Empty result tag received")]
    #[case::actually_empty(17, ResultOp::new_bool(""), &DEFAULT_STACK_ARRAY_LOWERING)]
    fn emit_result_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: ResultOp,
        #[case] array_lowering: &'static (impl ArrayLowering + Clone),
    ) {
        let pcg = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| {
            ceb.add_extension(ResultsCodegenExtension::new(array_lowering.clone()))
                .add_extension(array_lowering.codegen_extension())
                .add_prelude_extensions(pcg.clone())
                .add_default_int_extensions()
                .add_float_extensions()
        });
        let ext_op = op.to_extension_op().unwrap().into();
        let hugr = single_op_hugr(ext_op);
        check_emission!(hugr, llvm_ctx);
    }
}
