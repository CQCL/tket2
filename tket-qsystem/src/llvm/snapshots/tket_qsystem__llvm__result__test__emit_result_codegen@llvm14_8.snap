---
source: tket-qsystem/src/llvm/result.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

@res_test_arr_i.DFD30452.0 = private constant [25 x i8] c"\18USER:INTARR:test_arr_int"

define private void @_hl.main.1([10 x i64] %0) {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  %tag_len = load i8, i8* getelementptr inbounds ([25 x i8], [25 x i8]* @res_test_arr_i.DFD30452.0, i32 0, i32 0), align 1
  %tag_len2 = zext i8 %tag_len to i64
  %1 = alloca i64, i32 10, align 8
  %2 = bitcast i64* %1 to [10 x i64]*
  store [10 x i64] %0, [10 x i64]* %2, align 4
  %out_arr_alloca = alloca <{ i32, i32, i64*, i1* }>, align 8
  %x_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 0
  %y_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 1
  %arr_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 2
  %mask_ptr = getelementptr inbounds <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, i32 0, i32 3
  %3 = alloca i1, i32 10, align 1
  %4 = bitcast i1* %3 to [10 x i1]*
  store [10 x i1] zeroinitializer, [10 x i1]* %4, align 1
  store i32 10, i32* %x_ptr, align 4
  store i32 1, i32* %y_ptr, align 4
  store i64* %1, i64** %arr_ptr, align 8
  store i1* %3, i1** %mask_ptr, align 8
  %5 = load <{ i32, i32, i64*, i1* }>, <{ i32, i32, i64*, i1* }>* %out_arr_alloca, align 1
  call void @print_int_arr(i8* getelementptr inbounds ([25 x i8], [25 x i8]* @res_test_arr_i.DFD30452.0, i32 0, i32 0), i64 %tag_len2, <{ i32, i32, i64*, i1* }>* %out_arr_alloca)
  ret void
}

declare void @print_int_arr(i8*, i64, <{ i32, i32, i64*, i1* }>*)
